cpp/order/CubaticOrderParameter.cc

```diff
@@ -6,148 +6,175 @@
 }
 
-void CubaticOrderParameter::calcCubaticOrderParameter(float& cubatic_order_parameter, float* cubatic_tensor)
+//! Writeable index into array.
+float &tensor4::operator[](unsigned int index)
 {
-    tensor4<float> l_cubatic_tensor = tensor4<float>();
-    memcpy((void*) &l_cubatic_tensor.data, (void*) cubatic_tensor, sizeof(float) * 81);
-    tensor4<float> diff;
-    diff = m_global_tensor - l_cubatic_tensor;
-    cubatic_order_parameter = 1.0 - dot(diff, diff) / dot(l_cubatic_tensor, l_cubatic_tensor);
+    return data[index];
 }
 
-float CubaticOrderParameter::getCubaticOrderParameter()
+tensor4 tensor4::operator+=(const tensor4 &b)
 {
-    return m_cubatic_order_parameter;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        data[i] += b.data[i];
+    }
+    return *this;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getParticleCubaticOrderParameter()
+tensor4 tensor4::operator-(const tensor4 &b) const
 {
-    return m_particle_order_parameter;
+    tensor4 c;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        c.data[i] = data[i] - b.data[i];
+    }
+    return c;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getParticleTensor()
+tensor4 tensor4::operator*(const float &b) const
 {
-    return m_particle_tensor;
+    tensor4 c;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        c.data[i] = data[i] * b;
+    }
+    return c;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getGlobalTensor()
+void tensor4::copyToManagedArray(util::ManagedArray<float> &ma)
 {
-    memcpy(m_sp_global_tensor.get(), (void*) &m_global_tensor.data, sizeof(float) * 81);
-    return m_sp_global_tensor;
+    memcpy(ma.get(), (void*) &data, sizeof(float) * 81);
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getCubaticTensor()
+//! Complete tensor contraction.
+/*! This function is simply a sum-product over two tensors. For reference, see eq. 4.
+ *
+ *  \param a The first tensor.
+ *  \param a The second tensor.
+ */ 
+float dot(const tensor4& a, const tensor4& b)
 {
-    memcpy(m_sp_cubatic_tensor.get(), (void*) &m_cubatic_tensor.data, sizeof(float) * 81);
-    return m_sp_cubatic_tensor;
+    float c = 0;
+    for (unsigned int i = 0; i < 81; i++)
+    {
+        c += a.data[i] * b.data[i];
+    }
+    return c;
 }
 
-std::shared_ptr<float> CubaticOrderParameter::getGenR4Tensor()
+//! Generate the r4 tensor.
+/*! The r4 tensor is not a word used in the paper, but is a name introduced in
+ *  this code to refer to the second term in eqs. 27 in the paper. It is simply
+ *  a scaled sum of some delta function products. For convenience, its
+ *  calculation is performed in a single function.
+ */ 
+tensor4 genR4Tensor()
 {
-    memcpy(m_sp_gen_r4_tensor.get(), (void*) &m_gen_r4_tensor.data, sizeof(float) * 81);
-    return m_sp_gen_r4_tensor;
-}
+    // Construct the identity matrix to build the delta functions.
+    util::ManagedArray<float> identity({3, 3});
```

Comment:
This probably doesn't demand a change, but `identity` is a function in the standard library for C++20: https://en.cppreference.com/w/cpp/utility/functional/identity

(I only noticed because the syntax highlighting was different for that word.)

Reply:
Oh interesting. That's cool. But yeah not worth changing, even C++17 support isn't commonly or completely out of beta yet depending on which compiler you use.